chat-android/src/main/java/com/ably/chat/RoomOptions.kt

```diff
@@ -1,113 +1,187 @@
+
+@ChatDsl
+public class MutableRoomOptions : RoomOptions {
+    override var presence: MutablePresenceOptions? = null
+    override var typing: MutableTypingOptions? = null
+    override var reactions: MutableRoomReactionsOptions? = null
+    override var occupancy: MutableOccupancyOptions? = null
+}
+
+@ChatDsl
+public class MutablePresenceOptions : PresenceOptions {
+    override var enter: Boolean = true
+    override var subscribe: Boolean = true
+}
+
+@ChatDsl
+public class MutableTypingOptions : TypingOptions {
+    override var timeoutMs: Long = 5000
+}
+
+@ChatDsl
+public class MutableRoomReactionsOptions : RoomReactionsOptions
+
+@ChatDsl
+public class MutableOccupancyOptions : OccupancyOptions
+
+public fun buildRoomOptions(init: MutableRoomOptions.() -> Unit = {}): RoomOptions =
+    MutableRoomOptions().apply(init).asEquatable()
+
+public fun MutableRoomOptions.presence(init: MutablePresenceOptions.() -> Unit = {}) {
+    this.presence = MutablePresenceOptions().apply(init)
 }
 
+public fun MutableRoomOptions.typing(init: MutableTypingOptions.() -> Unit = {}) {
+    this.typing = MutableTypingOptions().apply(init)
+}
+
+public fun MutableRoomOptions.reactions(init: MutableRoomReactionsOptions.() -> Unit = {}) {
+    this.reactions = MutableRoomReactionsOptions().apply(init)
+}
+
+public fun MutableRoomOptions.occupancy(init: MutableOccupancyOptions.() -> Unit = {}) {
+    this.occupancy = MutableOccupancyOptions().apply(init)
+}
+
+internal data class EquatableRoomOptions(
+    override val presence: PresenceOptions? = null,
+    override val typing: TypingOptions? = null,
+    override val reactions: RoomReactionsOptions? = null,
+    override val occupancy: OccupancyOptions? = null,
+) : RoomOptions
+
+internal data class EquatablePresenceOptions(
+    override val enter: Boolean,
+    override val subscribe: Boolean,
+) : PresenceOptions
+
+internal data class EquatableTypingOptions(
+    override val timeoutMs: Long,
+) : TypingOptions
+
+internal data object EquatableRoomReactionsOptions : RoomReactionsOptions
+internal data object EquatableOccupancyOptions : OccupancyOptions
+
+internal fun MutableRoomOptions.asEquatable() = EquatableRoomOptions(
+    presence = presence?.asEquatable(),
+    typing = typing?.asEquatable(),
+    reactions = reactions?.let { EquatableRoomReactionsOptions },
+    occupancy = occupancy?.let { EquatableOccupancyOptions },
+)
+
+internal fun MutablePresenceOptions.asEquatable() = EquatablePresenceOptions(
+    enter = enter,
+    subscribe = subscribe,
+)
+
+internal fun MutableTypingOptions.asEquatable() = EquatableTypingOptions(
+    timeoutMs = timeoutMs,
+)
+
 /**
  * Throws AblyException for invalid room configuration.
  * Spec: CHA-RC2a
  */
 internal fun RoomOptions.validateRoomOptions(logger: Logger) {
     typing?.let {
-        if (typing.timeoutMs <= 0) {
-            logger.error("Typing timeout must be greater than 0, found ${typing.timeoutMs}")
+        if (it.timeoutMs <= 0) {
```

Comment:
Not a major change but `typing.timeoutMs` was more readable as first glance : )

Reply:
we have to replace it, because with interface kotlin complier doesn't know that property has a backing field

Reply:
we have to replace it, because with interface kotlin complier doesn't know that property has a backing field

Reply:
Okay, so does it create problem after compilation ..

Reply:
I can do 

```kotlin
typing?.let { typing -> 
    if (typing.timeoutMs <= 0) {}
}
```

but property shadowing I think even less readable

Reply:
In current case, it's easy to guess about `it` object, but in case the `lambda` is big enough, then it gets a bit less readable